### ISSUE TITLE
Is it a good idea that ... ?

### ISSUE THREAD

**Issue by khaing211** [angular/universal-starter#557]
Hi,

I have been researching the best way to implement AngularJS 5 in the following usecase:

1. The data provider is all backend i.e. either another service or a backend db requires credentials (that should not expose to the client in any way).
2. There won't be any exposed REST API to data provider i.e. there won't be /api/book/:id to get a particular book.
3. Since there is no REST, we fall back the normal HTTP GET and HTTP POST to fetch data and render the entire page.
4. Why is the need for Angular JS at all?  we would like to use Angular JS to render the page in the backend (i.e. SSR mode) and also handle animation in the browser.

I found the following code examples along with other guidance:

- Gotchas https://github.com/angular/universal#universal-gotchas (things not to do on server side for XYZ reason)
- Use isPlatformBrowser, isPlatformServer to determine the execution path. SSR AngularJS does not allow for dynamic dependencies due to AoT compilation. See example below. So we have to localize the use of the isPlatformBrowser, isPlatformServer  API to data provider and use TransferState API.

```
 import { PLATFORM_ID } from '@angular/core';
 import { isPlatformBrowser, isPlatformServer } from '@angular/common';
 
 constructor(@Inject(PLATFORM_ID) private platformId: Object) { ... }
 
 ngOnInit() {
   if (isPlatformBrowser(this.platformId)) {
      // Client only code.
      ...
   }
   if (isPlatformServer(this.platformId)) {
     // Server only code.
     ...
   }
 }
```

My ask is 

Since it is a standard practice to create a server NgModule along with browser NgModule, is it good idea to create a server Component along with a browser Component?

For example (GetBookDetails component):

- get-book.html
- getbook.server.ts (declare 'get-book.html')
- getbook.browser.ts (declare 'get-book.html')

**Comment by cmddavid** [comment]
I think you are trying to solve your back-end problem in the wrong area. Even though angular has server side rendering, it still essentially is a front-end framework. Trying to hide your api endpoints by only using them in your server.module could potentially work, but it would be a tough nut to crack. For instance, server side rendering is only triggered for the initial render, after that your application continues as a SPA. That means any route triggered after the initial load, won't have access to your server.module protected api endpoints.

**Comment by khaing211** [comment]
I agree with all you said, because that is the current target usecase Angular focus on. Constraint #3 above is artificial i.e. in the usecase of pagination or datatable, Ajax is, no doubt, a better implementation.

The reason I start to explore the path (described in my previous post) is that, under the current Angular guidance/standardize setup, Angular HttpClient on platform-server make another HTTP call back to fetch data inside Component. It does not make sense to me, because for every HTTP GET from browser, the server makes another HTTP GET to itself. This was one of the problems that motivate HttpCache /TransferState API where browser makes 2 HTTP GET (1. to get the index.html, 2. to get the data). platform-server simply moves the 2nd HTTP GET to be made from the server and does not make 2nd HTTP GET disappear.

Code Reference platform-server http.ts: https://github.com/angular/angular/blob/master/packages/platform-server/src/http.ts

My workaround around the issue above is to implement a HttpInterceptor to override with a local call to a database/backend data service rather than allowing a remote call (through some REST /api endpoint) to trigger a local call. Interceptor is passed to "providers" during bootstrap of ngExpressEngine.

Documentation: https://angular.io/api/common/http/HttpInterceptor

FYI, I have no knowledge if this issue has been raised.
